climada/hazard/base.py

```diff
@@ -1396,47 +1394,99 @@ def _read_att_excel(self, file_name, var_names):
-        Centroids of all hazards must have the same CRS.
+        All of the given hazards must be of the same type as self.
+        The centroids of all hazards must have the same CRS.
 
-        Tthe centroids of both hazards are combined together.
+        The centroids of all hazards are combined together.
         All raster centroids are converted to points and raster data
         is discarded.
 
-        Note: centroids of self is modified in place and raster information
+        Note: `self.centroids` is modified in place and centroid raster information (meta)
         is destroyed.
 
-        It is recommended to use the static method `hazard.concatenate_hazard`
-        which concatenates a list of hazards into a new object.
-
         Parameters
         ----------
-        hazard: climada.hazard.Hazard object
-            Hazard instance to append to self
+        haz_list : one or more climada.hazard.Hazard objects
+            Hazard instances to append to self
 
         Raises
         ------
         TypeError
 
         See Also
         --------
-        hazard.concat: concatenate 2 or more hazards
+        Hazard.concat: concatenate 2 or more hazards
         """
-        hazard._check_events()
-        if type(self) != type(hazard):
-            raise TypeError(f"hazard is of class {type(self)} which is "
-                            f" different from {type(hazard)}")
+        if len(haz_list) == 0:
+            return
+        haz_list = (self,) + haz_list
+        haz_list_nonempty = [haz for haz in haz_list if haz.size > 0]
 
-        if len(self.event_id) == 0: #if self is empty, replace self with hazard
-            self.__dict__ = hazard.__dict__
-        else:
-            self.__dict__ = Hazard.concat([self, hazard]).__dict__
+        for haz in haz_list:
+            haz._check_events()
+
+        # check type, unit, and attribute consistency among hazards
+        haz_types = {haz.tag.haz_type for haz in haz_list if haz.tag.haz_type != ''}
+        if len(haz_types) > 1:
+            raise ValueError(f"The given hazards are of different types: {haz_types}. "
+                             "The hazards are incompatible and cannot be concatenated.")
+        self.tag.haz_type = haz_types.pop()
+
+        haz_classes = {type(haz) for haz in haz_list}
+        if len(haz_classes) > 1:
+            raise TypeError(f"The given hazards are of different classes: {haz_classes}. "
+                            "The hazards are incompatible and cannot be concatenated.")
+
+        units = {haz.units for haz in haz_list if haz.units != ''}
+        if len(units) > 1:
+            raise ValueError(f"The hazards use different units: {units}. "
+                             "The hazards are incompatible and cannot be concatenated.")
+        elif len(units) == 0:
+            units = {''}
+        self.units = units.pop()
+
+        attributes = sorted(set.union(*[set(vars(haz).keys()) for haz in haz_list]))
```

Comment:
What is the sorted needed for here?

Reply:
This makes sure that we iterate over the attributes in a reproducible and consistent way in all of the following for-loops. From my experience, you should never iterate over a Python set. I think that shouldn't even be supported because it leads to inconsistent behavior when running the very same code twice on the same setup.

Reply:
I think here the order of iteration is irrelevant, so the sorting would not be strictly needed. But I do not mind it.

Reply:
The order of iteration is irrelevant if and only if everything works as expected. If you have a bug in your code, it can be extremely hard to debug due to inconsistent behavior. You obviously never had to deal with this kind of bug, otherwise you would mind. :D